st2common/st2common/persistence/base.py

```diff
@@ -112,12 +112,52 @@ def aggregate(cls, *args, **kwargs):
         return cls._get_impl().aggregate(*args, **kwargs)
 
     @classmethod
-    def add_or_update(cls, model_object, publish=True, dispatch_trigger=True):
+    def insert(cls, model_object, publish=True, dispatch_trigger=True,
+               log_not_unique_error_as_debug=False):
+        if model_object.id:
+            raise ValueError('id for object %s was unexpected.' % model_object)
+        try:
+            model_object = cls._get_impl().insert(model_object)
+        except NotUniqueError as e:
+            if log_not_unique_error_as_debug:
```

Comment:
add_or_update() makes a lazy programmer use the API incorrectly and will get us into errors. The semantics for insert, delete and update are well defined. I'd rather have them separate. I think we have had this discussion multiple times :P.